src/main/java/io/github/oliviercailloux/y2018/apartments/apartment/ApartmentFactory.java

```diff
@@ -0,0 +1,235 @@
+	 * 
+	 * @param nbApartment the number of apartments the list should contains
+	 * @return a list of random apartments of size nbApartment with real Address
+	 * @throws IOException if the Address API doesn't answer
+	 */
+	public static List<Apartment> generateRandomRealApartmentList(int nbApartment) throws IOException{
+		return generateRandomApartmentList(nbApartment,true);
+	}
+
+	/**
+	 * This function aims to generate a list of random apartments.
+	 *
+	 * @param nbApartment the number of apartments the list should contains
+	 * @return <i>List</i> a list of random apartments of size nbApartment
+	 * @throws IOException if the Address API doesn't answer
+	 */
+	private static List<Apartment> generateRandomApartmentList(int nbApartment,boolean realAddress) throws IOException {
+		if(nbApartment <=0) {
+			throw new IllegalArgumentException("You must indicate a number of apartments > 0");
+		}
+		List<Apartment> listApartment = new ArrayList<>();
+		for(int i = 0; i < nbApartment ; i++) {
+			listApartment.add(realAddress ? generateRandomRealApartment():generateRandomApartment());
+		}
+		return listApartment;
+	}
+
+	/**
+	 * This function simulates a random draw of a variable, being given its
+	 * expectation and its deviation.
+	 * 
+	 * @param mean gives the mathematical expectation of the variable
+	 * @param deviation is the standard deviation of the variable
+	 * @return an outcome of the randomized experiment
+	 */
+	private static double simulateRandomDraw(double mean, double deviation) {
+		double draw = Math.abs(rand.nextGaussian());
+		return deviation * draw + mean;
+	}
+
+	/**
+	 * Call an API which generates a random address.
+	 * This function aims at getting the random address generated.
+	 *
+	 * @param retry indicates how many time the function has been called (Retry System).
+	 * It will allow us to try escaping problems while contacting API (if it fails once, we try to call it
+	 * again until the number specified in the class attribute NB_MAX_RETRY).
+	 * @return the address generated.
+	 * @throws IOException if we cannot contact the API generator.
+	 */
+	private static String getRandomAddress(int retry) throws IOException {
+
+		String address = "";
+		//Code from https://www.developpez.net/forums/d1354479/java/general-java/recuperer-reponse-d-adresse-http/ 
+		try(InputStream is = new URL(URL_API_ADDRESS).openConnection().getInputStream()) { 
+			BufferedReader reader = new BufferedReader(new InputStreamReader(is, "utf-8"));   
+			StringBuilder builder = new StringBuilder(); 
+			for(String line = reader.readLine(); line != null; line = reader.readLine()) { 
+				builder.append(line + "\n"); 
+			} 
+			String bodyContent = builder.toString();
+			//End of code picking
+
+			return JsonConvert.getAddressFromJson(bodyContent); 
+
+		} catch (MalformedURLException e) {
+			LOGGER.error("Problem while contacting address generator API",e);
+		} catch (IOException e) {
+			if(retry < NB_MAX_RETRY) {
```

Comment:
Cette fct de retry n’a vraiment pas sa place ici, à la réflexion. On peut souhaiter modifier toutes sortes d’options concernant le moyen de requête réseau utilisé (timeout, retries, …), mais ce paramètre ici n’est pas le bon moyen de le faire. En outre, si la requête continue à échouer après plusieurs essais (ce qui sera généralement le cas si elle a échoué après le premier essai), le stack trace généré sera rendu peu lisible par les exceptions en cascade. (OPT)

Reply:
@oliviercailloux Remarque bien prise en compte ! 
Trop de changements à réfléchir pour le faire aujourd'hui, 
Ces modifications seront prises en compte dans la prochaine itération
cf : https://github.com/av1m/Apartments/issues/7